opm/input/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -1959,10 +1962,175 @@ std::vector<double> EclipseGrid::createDVector(const std::array<int,3>& dims, st
+    }
+    
+    void EclipseGrid::save_all_lgr_labels(const LgrCollection& lgr_input) {
+        all_lgr_labels.reserve(lgr_input.size()+1);
+        all_lgr_labels.push_back("GLOBAL");
+        for (std::size_t index = 0; index < lgr_input.size(); index++) {
+            all_lgr_labels.push_back(lgr_input.getLgr(index).NAME());
+        }    
+    }
+
+    void EclipseGrid::create_lgr_cells_tree(const LgrCollection& lgr_input) {                
+          auto IJK_global = [this](const auto& i_list, const auto& j_list, const auto& k_list){
+            if ((i_list.size() != j_list.size()) || (j_list.size() != k_list.size()) || (k_list.size() != i_list.size()) ){
+                 throw std::invalid_argument("Sizes are not compatible.");
+            }
+            std::vector<std::size_t> global_ind(i_list.size());
+            for (std::size_t index = 0; index < i_list.size(); index++) {              
+                global_ind[index] = this->getActiveIndex(i_list[index],j_list[index],k_list[index]);
+            }
+            return global_ind;
+        };
+        for (std::size_t index = 0; index < lgr_input.size(); index++) {
+            const auto lgr_cell = lgr_input.getLgr(index);
+            if (this->lgr_label.compare(lgr_cell.PARENT_NAME()) == 0)            {
+                lgr_grid = true;
+                auto [i_list, j_list, k_list] = lgr_cell.parent_cellsIJK();                
+                auto father_lgr_index = IJK_global(i_list, j_list, k_list);
+                lgr_children_cells.emplace_back(lgr_cell.NAME(), this->lgr_label, this->lgr_level, 
+                                                lgr_cell.NX(), lgr_cell.NY(), lgr_cell.NZ(), father_lgr_index);
+                
+                lgr_children_cells.back().create_lgr_cells_tree(lgr_input);              
+            }
+        }
+        std::sort(lgr_children_cells.begin(), lgr_children_cells.end(),[](const EclipseGridLGR& a, const EclipseGridLGR& b) {
+                return a.get_father_global()[0] < b.get_father_global()[0]; // Sort by another property
+        });
+        lgr_children_labels.reserve(lgr_children_cells.size());
+        for (auto lgr_cell : lgr_children_cells) {
+            lgr_children_labels.emplace_back(lgr_cell.lgr_label);
+        }
+        lgr_active_index.resize(lgr_children_cells.size(),0);
+    }
+
+    void EclipseGrid::init_lgr_cells_index(){
+        auto set_map_scalar = [&](const auto& vec, const auto& value){
+             num_lgr_children_cells[vec] = value;
+        };
+        auto set_vec_value = []( auto& vec, const auto& num_ref, const auto& data){
+            for (std::size_t index = 0; index < num_ref.size(); index++) {              
+                vec[num_ref[index]] = data;
+            }
+        };        
+        std::vector<std::size_t> lgr_level_numbering_counting(getNumActive(),1);
+        lgr_level_active_map.resize(getNumActive(),0);
+        for (auto cell:lgr_children_cells) {
```

Comment:
Please don't copy the `lgr_children_cells` sequence unless you need to.  Write the loop as `for (const auto&` unless there's a particular reason not to.